README.md
# prefect-census

Visit the full docs [here](https://PrefectHQ.github.io/prefect-census) to see additional examples and the API reference.

<p align="center">
    <a href="https://pypi.python.org/pypi/prefect-census/" alt="PyPI version">
        <img alt="PyPI" src="https://img.shields.io/pypi/v/prefect-census?color=0052FF&labelColor=090422"></a>
    <a href="https://github.com/PrefectHQ/prefect-census/" alt="Stars">
        <img src="https://img.shields.io/github/stars/PrefectHQ/prefect-census?color=0052FF&labelColor=090422" /></a>
    <a href="https://pepy.tech/badge/prefect-census/" alt="Downloads">
        <img src="https://img.shields.io/pypi/dm/prefect-census?color=0052FF&labelColor=090422" /></a>
    <a href="https://github.com/PrefectHQ/prefect-census/pulse" alt="Activity">
        <img src="https://img.shields.io/github/commit-activity/m/PrefectHQ/prefect-census?color=0052FF&labelColor=090422" /></a>
    <br>
    <a href="https://prefect-community.slack.com" alt="Slack">
        <img src="https://img.shields.io/badge/slack-join_community-red.svg?color=0052FF&labelColor=090422&logo=slack" /></a>
    <a href="https://discourse.prefect.io/" alt="Discourse">
        <img src="https://img.shields.io/badge/discourse-browse_forum-red.svg?color=0052FF&labelColor=090422&logo=discourse" /></a>
</p>

## Welcome!

This collection contains Prefect integrations for working with Census.

Census is an Operational Analytics platform that enables you to sync your trusted analytics data from your hub into operational tools that your business teams use on a daily basis.

Some things you can do with this collection out-of-the-box include:
- Trigger a Census sync run and wait for it to finish [(see how)](#use-a-censussync-job-block-to-run-a-sync-and-wait-for-completion)
- Call a custom endpoint with a `CensusClient` using your `CensusCredentials` [(see how)](#call-a-custom-endpoint)


For information on how to get started with Census, refer to the [Census docs](https://docs.getcensus.com/).

## Getting Started

### Python setup

Requires an installation of Python 3.7+.

We recommend using a Python virtual environment manager such as pipenv, conda or virtualenv.

These tasks are designed to work with Prefect 2.0. For more information about how to use Prefect, please refer to the [Prefect documentation](https://orion-docs.prefect.io/).

### Installation

Install `prefect-census` with `pip`:

```bash
pip install prefect-census
```

A list of available blocks in `prefect-census` and their setup instructions can be found [here](https://PrefectHQ.github.io/prefect-census/#blocks-catalog).

### Get a Census API Key

You will need a Census API key to be able to use the integrations in this collection. 

For directions for how to generate a Census API key, refer to the [Getting API Access](https://docs.getcensus.com/basics/api#getting-api-access) section of the Census docs.

Once you have a Census API key, you can configure a `CensusCredentials` block in the Prefect UI for use with the integrations in this collection. For information about how to configure a block in the Prefect UI, refer to the [Prefect docs](https://orion-docs.prefect.io/ui/blocks/).

### 💡  **NOTE**
> All [Census Syncs are incremental](https://docs.getcensus.com/basics/core-concept#sync-behaviors), which means they only capture records that are new or changed data since the previous sync.


### Write and run a flow


#### Use a `CensusSync` job block to run a sync and wait for completion

```python
from prefect import flow
from prefect_census import (
    CensusCredentials, CensusSync, run_census_sync
)

census_sync = CensusSync(
    credentials=CensusCredentials(api_key="my_api_key"),
    sync_id=42
)

@flow
def my_census_flow():
    # do some setup
    
    run_census_sync(census_sync)
    
    # do some cleanup

```

#### **Trigger Census sync run and wait for completion**:

Note that this works the same way as the `run_census_sync` above, however we recommend the `CensusSync` version as it implements the `JobBlock` interface that we intend to support.

```python
from prefect import flow

from prefect_census import CensusCredentials
from prefect_census.syncs import trigger_census_sync_run_and_wait_for_completion

@flow
def my_census_orchestrator():
    # do some setup
    creds = CensusCredentials(api_key="my_api_key")
    run_result = trigger_census_sync_run_and_wait_for_completion(
        credentials=creds,
        sync_id=42
    )
    # do some other things
```

#### **Get Census sync run info**:
```python
from prefect import flow

from prefect_census import CensusCredentials
from prefect_census.runs import get_census_sync_run_info

@flow
def get_sync_run_info_flow():
    credentials = CensusCredentials(api_key="my_api_key")

    return get_census_sync_run_info(
        credentials=credentials,
        run_id=42
    )

get_sync_run_info_flow()
```

#### **Call a custom endpoint**:
```python
from prefect import flow
from prefect_census import CensusCredentials
from prefect_census.client import CensusClient

@flow
def my_flow(sync_id):
    creds_block = CensusCredentials(api_key="my_api_key")

    client = CensusClient(
        api_key=creds_block.api_key.get_secret_value()    
    )
    response = client.call_endpoint(
        http_method="GET", path=f"/syncs/{sync_id}"
    )
    return response

my_flow(42)
```

For more tips on how to use tasks and flows in a Collection, check out [Using Collections](https://orion-docs.prefect.io/collections/usage/)!

## Resources

If you encounter any bugs while using `prefect-census`, feel free to open an issue in the [prefect-census](https://github.com/PrefectHQ/prefect-census) repository.

If you have any questions or issues while using `prefect-census`, you can find help in either the [Prefect Discourse forum](https://discourse.prefect.io/) or the [Prefect Slack community](https://prefect.io/slack).

Feel free to star or watch [`prefect-census`](https://github.com/PrefectHQ/prefect-census) for updates too!

## Contributing

If you'd like to help contribute to fix an issue or add a feature to `prefect-census`, please [propose changes through a pull request from a fork of the repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).

Here are the steps:
1. [Fork the repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository)
2. [Clone the forked repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#cloning-your-forked-repository)
3. Install the repository and its dependencies:
```
pip install -e ".[dev]"
```
4. Make desired changes
5. Add tests
6. Insert an entry to [CHANGELOG.md](https://github.com/PrefectHQ/prefect-census/blob/main/CHANGELOG.md)
7. Install `pre-commit` to perform quality checks prior to commit:
```
pre-commit install
```
8. `git commit`, `git push`, and create a pull request


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## 0.1.2

Released on July 20th, 2023.

### Added
- `CensusSyncRunStatus` enum extended to include the `queued` status value - [#28](https://github.com/PrefectHQ/prefect-census/pull/28)

## 0.1.1

Released on February 2nd, 2023.

### Added
- `CensusSync` job block and `CensusSyncRun` job run representing a Census sync job - [#15](https://github.com/PrefectHQ/prefect-census/pull/15)

### Fixed
- Logging within `trigger_census_sync_and_wait_for_completion` printed the builtin `id` instead of the Census sync ID - [#15](https://github.com/PrefectHQ/prefect-census/pull/15)

## 0.1.0

Released on October 26th, 2022.

### Added

- `trigger_census_sync` task - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)
- `get_census_sync_run_info` task - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)
- `CensusCredentials` block - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)
- `CensusClient` - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)
- `wait_census_sync_completion` flow - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)
- `trigger_census_sync_and_wait_for_completion` flow - [#10](https://github.com/PrefectHQ/prefect-census/pull/10)


tests/conftest.py
import pytest
from httpx import Response
from prefect.testing.utilities import prefect_test_harness


@pytest.fixture(autouse=True)
def prefect_db():
    with prefect_test_harness():
        yield


@pytest.fixture
def mock_successful_sync_calls(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        return_value=Response(
            200,
            json={"data": {"id": 5, "sync_run_id": 12345, "status": "completed"}},
        )
    )


@pytest.fixture
def mock_trigger_sync_calls(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/45/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"sync_run_id": 45}}))


@pytest.fixture
def mock_sync_not_found(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/46/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(404, json={"status": {"user_message": "Not found!"}}))


@pytest.fixture
def mock_successful_sync(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        return_value=Response(
            200,
            json={"data": {"id": 5, "sync_run_id": 12345, "status": "completed"}},
        )
    )


@pytest.fixture
def mock_successful_sync_with_wait(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        side_effect=[
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "queued"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "completed"}},
            ),
        ]
    )


@pytest.fixture
def mock_failed_sync_with_wait(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        side_effect=[
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "failed"}},
            ),
        ]
    )


@pytest.fixture
def mock_sync_failure_no_run_id(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "status": "completed"}}))


@pytest.fixture
def mock_cancelled_sync_with_wait(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        side_effect=[
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "cancelled"}},
            ),
        ]
    )


@pytest.fixture
def mock_sync_timed_out(respx_mock):
    respx_mock.post(
        "https://app.getcensus.com/api/v1/syncs/5/trigger",
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(return_value=Response(200, json={"data": {"id": 5, "sync_run_id": 12345}}))

    respx_mock.get(
        "https://app.getcensus.com/api/v1/sync_runs/12345",  # noqa
        headers={"Authorization": "Bearer my_api_key"},
    ).mock(
        side_effect=[
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
            Response(
                200,
                json={"data": {"id": 5, "sync_run_id": 12345, "status": "working"}},
            ),
        ]
    )


tests/test_runs.py
import pytest
from httpx import Response

from prefect_census.credentials import CensusCredentials
from prefect_census.runs import CensusGetSyncRunInfoFailed, get_census_sync_run_info


@pytest.fixture
def census_credentials():
    return CensusCredentials(api_key="my_api_key")


class TestGetCensusSyncRunInfo:
    async def test_get_census_sync_run_info(self, respx_mock, census_credentials):
        respx_mock.get(
            "https://app.getcensus.com/api/v1/sync_runs/42",  # noqa
            headers={"Authorization": "Bearer my_api_key"},
        ).mock(return_value=Response(200, json={"data": {"sync_run_id": 42}}))

        await get_census_sync_run_info.fn(credentials=census_credentials, run_id=42)

    async def test_get_nonexistent_run(self, respx_mock, census_credentials):
        respx_mock.get(
            "https://app.getcensus.com/api/v1/sync_runs/4",  # noqa
            headers={"Authorization": "Bearer my_api_key"},
        ).mock(
            return_value=Response(404, json={"status": {"message": "Not found!"}})
        )
        with pytest.raises(CensusGetSyncRunInfoFailed, match="Not found!"):
            await get_census_sync_run_info.fn(credentials=census_credentials, run_id=4)


prefect_census/flow.py


# import asyncio
# import os
# census_creds_block = CensusCredentials(api_key=os.environ["CENSUS_API_KEY"])

# c = census_creds_block.get_client()
# print("Call endpoint:", asyncio.run(c.call_endpoint(http_method="GET", path=f"/api/v1/syncs/38417")))

# print("Trigger_census_sync_run:", asyncio.run(trigger_census_sync_run.fn(census_creds_block, sync_id=38417)))
# print("Get census sync run info:", asyncio.run(get_census_sync_run_info.fn(census_creds_block, sync_id=38417)))

# from prefect import flow
    
# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import trigger_census_sync

# @flow
# def trigger_census_sync_run_flow():
#     credentials = CensusCredentials.load("my-census-credentials")
#     trigger_census_sync(credentials=credentials, sync_id=38417)
    
# trigger_census_sync_run_flow()

# import asyncio
# from prefect import flow
# from prefect_census.credentials import CensusCredentials

# credentials = CensusCredentials(api_key="my-api-key")

# @flow
# async def trigger_sync_run_flow(): 
#     async with credentials.get_client() as client:
#         await client.trigger_sync_run(sync_id=42)

# asyncio.run(trigger_sync_run_flow())



# from prefect import flow

# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import trigger_census_sync

# @flow
# def trigger_census_sync_run_flow():
#     credentials = CensusCredentials.load("my-census-credentials")
#     trigger_census_sync(credentials=credentials, sync_id=42)
    
# trigger_census_sync_run_flow()

# from prefect import flow
        
# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import get_census_sync_run_info

# @flow
# def get_sync_run_info_flow():
#     credentials = CensusCredentials(api_key="my_api")
    
#     return get_census_sync_run_info(
#         credentials=credentials,
#         sync_id=42
#     )

# get_sync_run_info_flow()

# from prefect import flow

# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import trigger_census_sync

# @flow
# def trigger_census_sync_flow():
#     credentials = CensusCredentials(api_key="my_api_key")
#     trigger_census_sync(credentials=credentials, sync_id=42)

# trigger_census_sync_flow()

# import os
# from prefect import flow
# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import trigger_census_sync_run_and_wait_for_completion

# @flow
# def my_flow():
#     creds = CensusCredentials(api_key=os.environ["CENSUS_API_KEY"])
#     run_result = trigger_census_sync_run_and_wait_for_completion(
#         credentials=creds,
#         sync_id=38417
#     )
#     print("run_result:", run_result)
# my_flow()

import os

# import asyncio
        
# from prefect_census.credentials import CensusCredentials
# from prefect_census.syncs import trigger_census_sync_run_and_wait_for_completion

# asyncio.run(
#     trigger_census_sync_run_and_wait_for_completion(
#         credentials=CensusCredentials(
#             api_key=os.environ["CENSUS_API_KEY"]
#         ),
#         sync_id=38417
#     )
# )


from prefect import flow

from prefect_census.credentials import CensusCredentials
from prefect_census.runs import get_census_sync_run_info

@flow
def get_sync_run_info_flow():
    credentials = CensusCredentials(api_key=os.environ["CENSUS_API_KEY"])
    
    return get_census_sync_run_info(
        credentials=credentials,
        run_id=69834024
    )

print(get_sync_run_info_flow())

prefect_census/credentials.py
"""Module containing credentials for interacting with Census."""
from prefect.blocks.abstract import CredentialsBlock
from pydantic import Field, SecretStr

from prefect_census.client import CensusClient


class CensusCredentials(CredentialsBlock):
    """
    Credentials block for credential use across Census tasks and flows.

    Attributes:
        api_key: API key to authenticate with the Census
            API. Refer to the [Census authentication docs](
            https://docs.getcensus.com/basics/api#getting-api-access)
            for retrieving the API key.

    Examples:
        Load stored Census credentials:
        ```python
        from prefect_census import CensusCredentials

        census_creds = CensusCredentials.load("BLOCK_NAME")
        ```

        Use CensusCredentials instance to trigger a sync run:
        ```python
        import asyncio
        from prefect import flow
        from prefect_census import CensusCredentials

        credentials = CensusCredentials(api_key="my_api_key")

        @flow
        async def trigger_sync_run_flow():
            async with credentials.get_client() as client:
                await client.trigger_sync_run(sync_id=42)

        asyncio.run(trigger_sync_run_flow())
        ```

        Load saved Census credentials within a flow:
        ```python
        from prefect import flow

        from prefect_census import CensusCredentials
        from prefect_census.syncs import trigger_census_sync

        @flow
        def trigger_census_sync_run_flow():
            credentials = CensusCredentials.load("my-census-credentials")
            trigger_census_sync(credentials=credentials, sync_id=42)

        trigger_census_sync_run_flow()
        ```
    """

    _block_type_name = "Census Credentials"
    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/3oznRx2UFkd2XyqNkEZpzB/4e0967a828aec5e2527cedadf8d24e8a/llmjpn8a0pgu8szjmnyi.webp?h=250"  # noqa

    api_key: SecretStr = Field(
        ..., title="API Key", description="API key to authenticate with the Census API."
    )

    def get_client(self) -> CensusClient:
        """
        Provides an authenticated client for working with the Census API.

        Returns:
            A authenticated Census API client
        """
        return CensusClient(api_key=self.api_key.get_secret_value())


prefect_census/syncs.py
"""Module containing tasks and flows for interacting with Census syncs."""
import asyncio
from typing import Any, Dict

from httpx import HTTPStatusError
from prefect import flow, task
from prefect.blocks.abstract import JobBlock, JobRun
from prefect.logging import get_run_logger
from prefect.utilities.asyncutils import sync_compatible
from pydantic import BaseModel, Field

from prefect_census.credentials import CensusCredentials
from prefect_census.runs import (
    CensusSyncRunCancelled,
    CensusSyncRunFailed,
    CensusSyncRunStatus,
    CensusSyncRunTimeout,
    wait_census_sync_completion,
)
from prefect_census.utils import extract_user_message


class CensusSyncTriggerFailed(RuntimeError):
    """Used to indicate sync triggered."""


class CensusSyncResult(BaseModel):
    """The results for a Census Sync Run."""

    final_status: CensusSyncRunStatus
    run_data: Dict[str, Any]


@task(
    name="Trigger Census sync run",
    description="Triggers a Census sync run for the sync with the given sync_id.",
    retries=3,
    retry_delay_seconds=10,
)
async def trigger_census_sync(
    credentials: CensusCredentials, sync_id: int, force_full_sync: bool = False
) -> int:
    """
    A task to trigger a Census sync run.

    Args:
        credentials: Credentials for authenticating with Census.
        sync_id: The ID of the sync to trigger.
        force_full_sync: If `True`, a full sync will be triggered.

    Returns:
        The ID of the triggered sync run.

    Examples:
        Trigger a Census sync run:
        ```python
        from prefect import flow

        from prefect_census import CensusCredentials
        from prefect_census.syncs import trigger_census_sync

        @flow
        def trigger_census_sync_flow():
            credentials = CensusCredentials(api_key="my_api_key")
            trigger_census_sync(credentials=credentials, sync_id=42)

        trigger_census_sync_flow()
        ```
    """
    logger = get_run_logger()

    logger.info(f"Triggering Census sync run for sync with ID {sync_id}")
    try:
        async with credentials.get_client() as client:
            response = await client.trigger_sync_run(
                sync_id=sync_id, force_full_sync=force_full_sync
            )
    except HTTPStatusError as e:
        raise CensusSyncTriggerFailed(extract_user_message(e)) from e

    run_data = response.json()["data"]

    if "sync_run_id" in run_data:
        logger.info(
            f"Census sync run successfully triggered for sync with ID {sync_id}. "
            "You can view the status of this sync run at "
            f"https://app.getcensus.com/sync/{sync_id}/sync-history"
        )

    return run_data["sync_run_id"]


@flow(
    name="Trigger Census sync run and wait for completion",
    description="Triggers a Census sync run and waits for the"
    "triggered run to complete.",
)
async def trigger_census_sync_run_and_wait_for_completion(
    credentials: CensusCredentials,
    sync_id: int,
    force_full_sync: bool = False,
    max_wait_seconds: int = 900,
    poll_frequency_seconds: int = 10,
) -> Dict[str, Any]:
    """
    Flow that triggers a sync run and waits for the triggered run to complete.

    Args:
        credentials: Credentials for authenticating with Census.
        sync_id: The ID of the sync to trigger.
        force_full_sync: If `True`, a full sync will be triggered.
        max_wait_seconds: Maximum number of seconds to wait for sync to complete
        poll_frequency_seconds: Number of seconds to wait in between checks for run completion.

    Raises:
        CensusSyncRunCancelled: The triggered Census sync run was cancelled.
        CensusSyncRunFailed: The triggered Census sync run failed.
        RuntimeError: The triggered Census sync run ended in an unexpected state.

    Returns:
        The final run data returned by the Census API as dict with the following shape:
            ```
            {
                "id": 94,
                "sync_id": 52,
                "source_record_count": 1,
                "records_processed": 1,
                "records_updated": 1,
                "records_failed": 0,
                "records_invalid": 0,
                "created_at": "2021-10-20T02:51:07.546Z",
                "updated_at": "2021-10-20T02:52:29.236Z",
                "completed_at": "2021-10-20T02:52:29.234Z",
                "scheduled_execution_time": null,
                "error_code": null,
                "error_message": null,
                "error_detail": null,
                "status": "completed",
                "canceled": false,
                "full_sync": true,
                "sync_trigger_reason": {
                    "ui_tag": "Manual",
                    "ui_detail": "Manually triggered by test@getcensus.com"
                }
            }
            ```

    Examples:
        Trigger a Census sync using CensusCredentials instance and wait
        for completion as a standalone flow:
        ```python
        import asyncio

        from prefect_census import CensusCredentials
        from prefect_census.syncs import trigger_census_sync_run_and_wait_for_completion

        asyncio.run(
            trigger_census_sync_run_and_wait_for_completion(
                credentials=CensusCredentials(
                    api_key="my_api_key"
                ),
                sync_id=42
            )
        )
        ```

        Trigger a Census sync and wait for completion as a subflow:
        ```python
        from prefect import flow

        from prefect_census import CensusCredentials
        from prefect_census.syncs import trigger_census_sync_run_and_wait_for_completion

        @flow
        def my_flow():
            ...
            creds = CensusCredentials(api_key="my_api_key")
            run_result = trigger_census_sync_run_and_wait_for_completion(
                credentials=creds,
                sync_id=42
            )
            ...

        my_flow()
        ```
    """  # noqa
    logger = get_run_logger()

    triggered_run_data_future = await trigger_census_sync.submit(
        credentials=credentials, sync_id=sync_id, force_full_sync=force_full_sync
    )

    run_id = await triggered_run_data_future.result()
    if run_id is None:
        raise RuntimeError("Unable to determine run ID for triggered sync.")

    final_run_status, run_data = await wait_census_sync_completion(
        run_id=run_id,
        credentials=credentials,
        max_wait_seconds=max_wait_seconds,
        poll_frequency_seconds=poll_frequency_seconds,
    )

    if final_run_status == CensusSyncRunStatus.COMPLETED:
        logger.info(
            "Census sync run with ID %s completed successfully!",
            run_id,
        )
        return run_data

    elif final_run_status == CensusSyncRunStatus.CANCELLED:
        raise CensusSyncRunCancelled(
            f"Triggered sync run with ID {run_id} was cancelled."
        )
    elif final_run_status == CensusSyncRunStatus.FAILED:
        raise CensusSyncRunFailed(f"Triggered sync run with ID: {run_id} failed.")
    else:
        raise RuntimeError(
            f"Triggered sync run with ID: {run_id} ended with unexpected"
            f"status {final_run_status}"
        )


class CensusSync(JobBlock):
    """
    A Job Block for triggering a Census sync run and waiting for completion.

    Attributes:
        credentials: Credentials for authenticating with Census.
        sync_id: The ID of the sync to trigger.
        force_full_sync: If `True`, a full sync will be triggered.
        max_wait_seconds: Maximum number of seconds to wait for sync to complete.
        poll_frequency_seconds: Number of seconds to wait in between checks
            for run completion.

    Example:
        Trigger a Census sync and wait for completion as a subflow:
        ```python
        from prefect import flow
        from prefect_census import CensusSync, run_census_sync

        @flow
        def my_census_flow():
            census_sync = CensusSync.load("BLOCK_NAME")

            # do some setup

            run_census_sync(census_sync)

            # do some cleanup
        ```
    """

    credentials: CensusCredentials = Field(
        default=...,
        description="Credentials for authenticating with Census.",
    )

    sync_id: int = Field(
        default=...,
        description="The ID of the sync to trigger.",
    )

    force_full_sync: bool = Field(
        default=False,
        description="If `True`, a full sync will be triggered.",
    )

    max_wait_seconds: int = Field(
        default=3600,
        description="The maximum number of seconds to wait for the sync to complete.",
    )

    poll_frequency_seconds: int = Field(
        default=5,
        description="Number of seconds to wait between sync status checks.",
    )

    _block_type_name = "Census Sync"
    _description = "Runs a Census sync"
    _documentation_url = "https://prefecthq.github.io/prefect-census/syncs/"
    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/3oznRx2UFkd2XyqNkEZpzB/4e0967a828aec5e2527cedadf8d24e8a/llmjpn8a0pgu8szjmnyi.webp?h=250"  # noqa

    @sync_compatible
    async def trigger(self) -> "CensusSyncRun":
        """
        Trigger a Census sync run.

        Returns:
            A CensusSyncRun instance representing the triggered sync run.
        """
        self.logger.info(f"Triggering Census sync run for sync with ID {self.sync_id}")
        try:
            async with self.credentials.get_client() as client:
                response = await client.trigger_sync_run(
                    sync_id=self.sync_id, force_full_sync=self.force_full_sync
                )
        except HTTPStatusError as e:
            raise CensusSyncTriggerFailed(extract_user_message(e)) from e

        run_data = response.json()["data"]

        if "sync_run_id" in run_data:
            self.logger.info(
                f"Census sync with ID: {self.sync_id} successfully triggered. "
                "You can view the status of this sync run at "
                f"https://app.getcensus.com/sync/{self.sync_id}/sync-history"
            )

        run_id = run_data["sync_run_id"]
        if run_id is None:
            raise RuntimeError("Unable to determine run ID for triggered sync.")

        return CensusSyncRun(census_sync=self, run_id=run_id)


class CensusSyncRun(JobRun):
    """
    A Job Run representing a Census sync run.

    Attributes:
        sync: The CensusSync instance that triggered this run.
        run_id: The ID of the Census sync run.
        run_data: The data returned by the Census API for this run.
        status: The status of the Census sync run.
    """

    def __init__(self, census_sync, run_id) -> None:
        self.sync = census_sync
        self.run_id = run_id
        self.run_data = None
        self.status = None

    @sync_compatible
    async def wait_for_completion(self):
        """Wait for the Census sync run to complete."""
        seconds_waited_for_run_completion = 0
        _last_status = None
        while seconds_waited_for_run_completion <= self.sync.max_wait_seconds:
            try:
                async with self.sync.credentials.get_client() as client:
                    response = await client.get_run_info(self.run_id)
            except HTTPStatusError as e:
                raise RuntimeError(extract_user_message(e)) from e

            run_data = response.json()["data"]
            self.status = run_data.get("status")

            if CensusSyncRunStatus.is_terminal_status_code(self.status):
                self.run_data = run_data
                if self.status == CensusSyncRunStatus.COMPLETED.value:
                    self.logger.info(
                        "Census sync run with ID %s completed successfully!",
                        self.run_id,
                    )
                    return

                elif self.status == CensusSyncRunStatus.CANCELLED.value:
                    raise CensusSyncRunCancelled(
                        f"Triggered sync run with ID {self.run_id} was cancelled."
                    )
                elif self.status == CensusSyncRunStatus.FAILED.value:
                    raise CensusSyncRunFailed(
                        f"Triggered sync run with ID {self.run_id} has failed."
                    )
                else:
                    raise RuntimeError(
                        f"Sync run with ID: {self.run_id} ended with unexpected "
                        f"status {self.status}"
                    )
            if self.status != _last_status:
                self.logger.info(
                    "Census sync run with ID %i has status %s.",
                    self.run_id,
                    CensusSyncRunStatus(self.status).name,
                )
                _last_status = self.status

            await asyncio.sleep(self.sync.poll_frequency_seconds)
            seconds_waited_for_run_completion += self.sync.poll_frequency_seconds

        raise CensusSyncRunTimeout(
            f"Timeout of {self.sync.max_wait_seconds} seconds exceeded while "
            f"waiting for sync run with ID {self.run_id} to complete."
        )

    @sync_compatible
    async def fetch_result(self) -> CensusSyncResult:
        """Fetch the result of the Census sync run.

        Returns:
            A CensusSyncResult instance representing the result of the sync run.
        """
        return CensusSyncResult(
            final_status=CensusSyncRunStatus(self.status),
            run_data=self.run_data,
        )
